File: TESTBENCH_LIBs_Cpp/include/Server_Assembly/Execute.cpp
#include "pch.h"
#include <cstddef>
#include <thread>

#include "LIB_Server_Assembly.h"

class Avril_FSD::Execute_Control* ptr_Execute_Control = NULL;
std::thread* ptr_Thread_WithCoreId[4] = { NULL, NULL, NULL, NULL };//NUMBER OF CORES

Avril_FSD::Execute::Execute(Global* ptr_Global, __int8 number_Implemented_Cores)
{

}

Avril_FSD::Execute::~Execute()
{
    delete ptr_Execute_Control;
    for (int index = 0; index < 4; index++)
    {
        delete ptr_Thread_WithCoreId[index];
    }//NUMBER OF CORES
}

void Avril_FSD::Execute::Initialise()
{
    Avril_FSD::Framework_Server::Get_Server_Assembly()->Get_Algorithms()->Initialise(
        Avril_FSD::Framework_Server::Get_Server_Assembly()->Get_Global()->Get_NumCores()
    );
}

void Avril_FSD::Execute::Initialise_Control(
    __int8 number_Implemented_Cores,
    Avril_FSD::Global* ptr_Global
)
{
    Set_Control_Of_Execute(new class Avril_FSD::Execute_Control(number_Implemented_Cores));
    while (Get_Control_Of_Execute() == NULL) { /* wait untill created */ }
}
void Avril_FSD::Execute::Initialise_Libraries()
{
    class Avril_FSD::Framework_Server* framework = Avril_FSD::LIB_ServerLibrary::Initialise_Server_Assembly();
    class Avril_FSD::Framework_Server* framework = Avril_FSD::LIB_ServerLibrary::Initialise_Server_Assembly();
}
void Avril_FSD::Execute::Initialise_Threads()
{
    for (__int8 coreId = 0; coreId < Avril_FSD::Framework_Server::Get_Server_Assembly()->Get_Global()->Get_NumCores(); coreId++)
    {
        ptr_Thread_WithCoreId[coreId] = new std::thread(
            Avril_FSD::Framework_Server::Get_Server_Assembly()->Get_Algorithms()->Get_Concurrent(coreId)->Thread_Concurrency,
            coreId,
            Avril_FSD::Framework_Server::Get_Server_Assembly()->Get_Global()->Get_NumCores()
        );
    }
}

Avril_FSD::Execute_Control* Avril_FSD::Execute::Get_Control_Of_Execute()
{
    return ptr_Execute_Control;
}

void Avril_FSD::Execute::Set_Control_Of_Execute(Execute_Control* execute_Control)
{
    ptr_Execute_Control = execute_Control;
}

File: APP_Server/engine/Execute.cs
using System.Threading;

namespace Avril_FSD_ServerAssembly
{ 
    public class Execute
    {
        Avril_FSD_ServerAssembly.Execute_Control execute_Control = null;
        Thread input_Listen = null;
        Thread output_Send = null;

        public Execute()
        {
            input_Listen = null;
            output_Send = null;
        }
        public void Initialise()
        {
            Avril_FSD_ServerAssembly.Framework.GetGameServer().GetAlgorithms().Initialise(Avril_FSD_ServerAssembly.Framework.GetGameServer().GetGlobal().Get_NumCores());
        }

        public void Initialise_Control(
            byte numberOfCores,
            Global global
        )
        {
            execute_Control = new Avril_FSD_ServerAssembly.Execute_Control(numberOfCores);
            while (execute_Control == null) { /* Wait while is created */ }
        }

        public void Initialise_Threads()
        {
            input_Listen = new Thread(Avril_FSD_ServerAssembly.Framework.GetGameServer().GetAlgorithms().GetIO_ListenRespond().Thread_Input_Listen);
            input_Listen.Start();

            output_Send = new Thread(Avril_FSD_ServerAssembly.Framework.GetGameServer().GetAlgorithms().GetIO_ListenRespond().Thread_Output_Send);
            output_Send.Start();
        }

        public void Create_And_Run_Graphics()
        {
            using (Avril_FSD_ServerAssembly.Game_Instance gameInstance = new Avril_FSD_ServerAssembly.Game_Instance())
            {
                gameInstance.Run(Avril_FSD_ServerAssembly.Framework.GetGameServer().GetData().GetSettings().Get_refreshRate());
            }
        }

        public Execute_Control GetExecute_Control()
        {
            return execute_Control;
        }
    }
}